CHANGELOG.md

```diff
@@ -2,6 +2,110 @@
 
 This contains only the most important and/or user-facing changes; for a full changelog, see the commit history.
 
+## [2.0.0](https://github.com/ably/ably-js/tree/2.0.0) (date TBD)
+
+The 2.0.0 release introduces a number of new features and QoL improvements, including a new way to remove bloat and reduce the bundle size of your ably-js client, first-class support for Promises, a more idiomatic approach to using ably-js' React Hooks, enhancements to TypeScript typings, and more.
+
+Below is a list of all new features, APIs, deprecations, and breaking changes.
+Read [ably-js v2 migration guide](link TBD) for more information.
+
+### Modular variant of the library
+
+An ESM variant of the library is now available for browsers (but not for Node.js) at `ably/modular`. This modular variant of the library supports tree shaking, allowing for a reduction in the Ably bundle size within your application. It can also be used by Web Workers.
+
+### First-class support for Promises
+
+The exports `ably/callbacks` and `ably/promises` have been removed. The callbacks API has been entirely removed, and now only promises are supported. Calls like `new Ably.Realtime.Promise` should be replaced with `new Ably.Realtime`.
+
+### React Hooks
+
+React hooks, exported at `ably/react`, now rely on the `ChannelProvider` component to define the channels you wish to use and the options for them.
+
+### TypeScript typings
+
+The Types namespace has been removed. All types are now exported at the top level.
+
+### Browser and Web Worker bundles
+
+- The browser bundle now relies on the native Web Crypto API instead of CryptoJS. The `ably/build/ably.noencryption` bundle has been removed, as it is no longer necessary.
+- The separate bundle for web workers, `ably/build/ably-webworker`, has been removed. You can now import from `ably` in web workers.
+
+### Supported platforms changes
+
+- Support for Internet Explorer has been dropped.
+- Support for Node.js versions lower than 16 has been dropped. The supported Node.js versions are now 16, 18, and 20.
+- The minimum supported versions for major browsers are: Chrome 58, Firefox 52, Edge 79, Safari 11, and Opera 45.
+
+___
+
+### Breaking Changes
```

Comment:
In general, I don't think we should "hide" information if it can be easily displayed. We have split the changelog into sections so that users who are interested in the high-level view can read only the first part, and if they want more specific information, they can read about merged PRs.

> refactorings: I'm not sure why users of the library would be interested in these internal changes

Even though some of the refactorings made are breaking changes, I would agree that it might not be necessary to list them separately again in the "Code Refactoring" section - one mention in the "Breaking Changes" section probably is enough.

> features and breaking changes

I think it's beneficial to list them all in a compressed format, one by one. The high-level overview above does not go into the specifics of how or why some things work, and those PRs provide quick access to that information. The migration guide, on the other hand, is very low-level, and there is no way to have a really quick look at it and get a feel for the required changes to migrate to v2.

Reply:
> In general, I don't think we should "hide" information if it can be easily displayed

I mean, if you think it's beneficial to include these lists of PRs in the changelog, I guess we can, but I would suggest that they be de-emphasised, perhaps by putting them inside `<details>` tags and making it clear that what follows is a list of PRs included in the changelog.

As it currently stands, the list of PRs looks like something that the user is _expected_ to read through in order to understand what's changed in v2 (especially since this changelog entry is likely to be a user’s first point of contact with v2). I can't speak for all users, but I certainly would not enjoy being expected to read through a bunch of PRs upon upgrading to a new version of a library that I use.

Bear in mind the intended audience of a pull request; it's not aimed at users, but rather at other developers in the team, who have knowledge about the context in which the PR is being created which a regular user wouldn’t have (or be expected to have). Furthermore, a single PR doesn't necessarily represent a single user-facing change; sometimes a change is split up into several PRs (possibly being performed iteratively, not necessarily getting it right the first time; software development is messy), and sometimes a PR contains a bunch of unrelated changes, some of which are relevant to users, some of which aren't (e.g. refactors). And, then, once a PR is created, it goes through a bunch of feedback and changes, meaning that the final state of a PR might not even fully reflect the changes described in its title or description, so to fully understand the impact of a PR a user would need to then look at all this stuff too.

> I think it's beneficial to list them all in a compressed format, one by one. The high-level overview above does not go into the specifics of how or why some things work, and those PRs provide quick access to that information. 

The fact the list of PRs appears concise doesn't necessarily imply that it's understandable to users.

> The migration guide, on the other hand, is very low-level

I’m not sure what you mean by "low-level" here. It's written in prose and explicitly aimed at users of the library. How is this more "low-level" than PRs which, ultimately, show the _changes to the library’s implementation_?

Reply:
So, to summarise — I think that we can include a list of PRs as "bonus" information for particularly interested users to view. But I think that any information that we consider important to communicate in this changelog should be communicated as prose that's explicitly targeted at our users.

Reply:
> some of the refactorings made are breaking changes

What's your definition of a "refactoring" here? My understanding of a refactoring is that it's a change that has no user-facing impact, i.e. it could never be a breaking change.

Reply:
> by putting them inside `details` tags and making it clear that what follows is a list of PRs included in the release.

> As it currently stands, the list of PRs looks like something that the user is expected to read through in order to understand what's changed in v2

> so to fully understand the impact of a PR a user would need to then look at all this stuff too.

Agree, I never intended for users to go through each PR and figure out what's changed - this is why we have a thorough migration guide that just tells you what to do. I've put PRs in the details tag as per your suggestion.

> But I think that any information that we consider important to communicate in this changelog should be communicated as prose that's explicitly targeted at our users.

Agree, this is how changelog looks like after hiding PRs in the details. Will also change wording in couple of sections to be less code-specific (like https://github.com/ably/ably-js/pull/1656#discussion_r1530506245 below)

Reply:
> What's your definition of a "refactoring" here? My understanding of a refactoring is that it's a change that has no user-facing impact, i.e. it could never be a breaking change.

Probably not the best choice of a word here. Refactorings are changes that restructure the code in some way, but do not alter its functionality. However even a change like this can be a breaking change: changing a name of a prop in some object which is part of the public API - looks like refactoring, but is still a breaking change. On this note, you could argue that changing public API **is** changing functionality in the library context - maybe, but it does not feel like a "feature" or a "bug fix" either. So I just put them under refactorings section.

Now, based on what was discussed above, I do feel now that there is no need to be so verbose when listing those PRs.
Currently, all user-facing PRs are split into three sections: "features", "bug fixes" and "refactorings", and then if any PR is a breaking change, it is then **also** listed in the breaking changes section.

I've changed that to list each PR only once, either in "breaking changes", "features" or "bug fixes" and removed "refactorings" section altogether.

Reply:
Yeah — changing a property name in the public API is a breaking change, not a refactoring. The new headings that you've suggested seem sensible to me.